FaustCode/owl.cpp

```diff
@@ -347,6 +360,194 @@ class OwlButton : public OwlParameterBase {
+
+    /**
+     * Get the channels and length values of the given sound resource.
+     *
+     * @param path_name - the name of the file, or sound resource identified this way
+     * @param channels - the channels value to be filled with the sound resource number of channels
+     * @param length - the length value to be filled with the sound resource length in frames
+     *
+     */
+    void getParamsFile(const char* path_name, int& channels, int& length) {
+        WaveResourceReader reader(path_name);
+        channels = 1;
+        length = 0;
+
+        if (reader.loadWaveHeader()){
+            channels = reader.fWave->num_channels;
+            length = (reader.fWave->subchunk_2_size * 8) / (reader.fWave->num_channels * reader.fWave->bits_per_sample);
+        }
+    }
+
+    /**
+     * Read one sound resource and fill the 'soundfile' structure accordingly
+     *
+     * @param soundfile - the soundfile to be filled
+     * @param path_name - the name of the file, or sound resource identified this way
+     * @param part - the part number to be filled in the soundfile
+     * @param offset - the offset value to be incremented with the actual sound resource length in frames
+     * @param max_chan - the maximum number of mono channels to fill
+     *
+     */
+    void readFile(Soundfile* soundfile, const char* path_name, int part, int& offset, int max_chan) {
+        WaveResourceReader reader(path_name);
+        if (!reader.loadWaveHeader()){
+            return;
+        }
+        reader.loadWave();
+
+        soundfile->fLength[part] = (reader.fWave->subchunk_2_size * 8) / (reader.fWave->num_channels * reader.fWave->bits_per_sample);
+        soundfile->fSR[part] = reader.fWave->sample_rate;
+        soundfile->fOffset[part] = offset;
+
+        // Audio frames have to be written for each chan
+        switch (reader.fWave->bits_per_sample){
+        case 16:
+            // 16 bit samples are signed ints
+            for (int sample = 0; sample < soundfile->fLength[part]; sample++) {
+                float factor = 1.f/32767.f;
+                int16_t* frame = (int16_t*)&reader.fWave->data[reader.fWave->block_align * sample];
+                for (int chan = 0; chan < reader.fWave->num_channels; chan++) {
+                    soundfile->fBuffers[chan][offset + sample] = frame[chan] * factor;
+                }
+            }
+            break;
+        case 8:
+            // 8 bit samples are unsigned ints
+            for (int sample = 0; sample < soundfile->fLength[part]; sample++) {
+                float factor = 1.f/255.f;
+                uint8_t* frame = (uint8_t*)&reader.fWave->data[reader.fWave->block_align * sample];
+                for (int chan = 0; chan < reader.fWave->num_channels; chan++) {
+                    soundfile->fBuffers[chan][offset + sample] = frame[chan] * factor - 0.5f;
+                }
+            }
+            break;
+        default:
+            debugMessage("Unsupported format");
```

Comment:
should this be: `error(PROGRAM_ERROR_STATUS, "Unsupported resource format")`
ie is it a recoverable error? If not use `error()`.

Reply:
Yes, it's recoverable, because FAUST mandates an empty buffer to be used in such cases. So if we can't parse a file or load resource, the buffer would point to that global empty buffer.